lib/translator/status.cc

```diff
@@ -0,0 +1,243 @@
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kMediumError;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedPowerLoss:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      result.asc = scsi_defs::AdditionalSenseCode::kWarningPowerLossExpected;
+      result.ascq =
+          scsi_defs::AdditionalSenseCodeQualifier::kWarningPowerLossExpected;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kInternalDeviceError:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kHardwareError;
+      result.asc = scsi_defs::AdditionalSenseCode::kInternalTargetFailure;
+      result.ascq =
+          scsi_defs::AdditionalSenseCodeQualifier::kInternalTargetFailure;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedByRequest:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedSqDeletion:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedFailedFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedMissingFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kInvalidNamespaceOrFormat:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc =
+          scsi_defs::AdditionalSenseCode::kAccessDeniedInvalidLuIdentifier;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kAccessDeniedInvalidLuIdentifier;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kLbaOutOfRange:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc = scsi_defs::AdditionalSenseCode::kLbaOutOfRange;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::kLbaOutOfRange;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kNamespaceNotReady:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kNotReady;
+      result.asc = scsi_defs::AdditionalSenseCode::
+          kLogicalUnitNotReadyCauseNotReportable;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kLogicalUnitNotReadyCauseNotReportable;
+      break;
+    default:
+      DebugLog("No SCSI translation for NVMe status with code %#x",
+               static_cast<uint8_t>(sc));
```

Comment:
The caller has no way to detect this case (and all the similar fallbacks below), so we need to either fill in the output with *something* or add a return value indicating whether the result was written which the next layer up will have to do something with.

From a high level viewpoint, if we're translating NVMe to SCSI, we're going to have to come up with *some* response no matter what strange condition arises, otherwise the SCSI user won't be able to make progress.